### ISSUE TITLE
Updating macOS release manifest susceptible to race conditions

### ISSUE THREAD

**Issue by christopherfujino** [flutter/flutter#100390]
Since https://flutter.googlesource.com/infra/+/2b10a1dc69802744cd4d8600e13f4897bae1f83b%5E%21/#F3, we have started building release packages for both arm64 and x64 on macOS. These are packages are created by the script: https://github.com/flutter/flutter/blob/master/dev/bots/prepare_package.dart. What this script roughly does is:

1. check out the flutter/flutter repo at the desired release branch and revision
2. run `flutter precache --all-platforms`
3. compress the SDK directory
4. upload it to `storage.googleapis.com/flutter_infra_release/releases/$CHANNEL/$HOST_PLATFORM/$ARCHIVE`
5. download the manifest JSON (which gets consumed by the website) from `storage.googleapis.com/flutter_infra_release/releases/releases_${HOST_PLATFORM}.json`
6. update the manifest with the current release
7. re-upload the updated manifest file.

The problem is that if we are scheduling concurrent macOS packaging builds (one for ARM64, one for x64), both builds could download the same manifest in step 5, but then whichever build reaches step 7 last could overwrite the changes from the other.

To solve this, I think we need to either:

a) have the two architectures use separate manifest files, and update the website to merge them when rendering the website; or
b) implement some kind of locking and polling setup. However, I'm not sure if even this would not be susceptible to race conditions, unless GCS provides a way to atomically set a lock if it is not already locked, and fail otherwise.

**Comment by christopherfujino** [comment]
cc @XilaiZhang, @CaseyHillers, @itsjustkevin

**Comment by CaseyHillers** [comment]
Is this something that would be true for all the packaging builders, such as linux and windows?

We could use [GCS Bucket Locking](https://cloud.google.com/storage/docs/using-bucket-lock) for (B)

**Comment by christopherfujino** [comment]
> Is this something that would be true for all the packaging builders, such as linux and windows?
> 
> We could use [GCS Bucket Locking](https://cloud.google.com/storage/docs/using-bucket-lock) for (B)

I don't believe we build linux and windows for multiple architectures...yet.

**Comment by christopherfujino** [comment]
> Is this something that would be true for all the packaging builders, such as linux and windows?
> 
> We could use [GCS Bucket Locking](https://cloud.google.com/storage/docs/using-bucket-lock) for (B)

Unfortunately GCS bucket locking does not put a mutex lock on a bucket, but rather prevents future changes to the bucket's retention policy.

Per this SO answer from 2019, GCS simply does not support locking: https://stackoverflow.com/questions/54768768/gcs-file-locking-query.

It sounds like we could use generation numbers and request preconditions to cause a write to fail if the current generation number isn't the same one as when we first read the file: https://cloud.google.com/storage/docs/metadata#generation-number

**Comment by christopherfujino** [comment]
Looks like when we write, we can add the precondition like so:

```
gcloud storage cp OBJECT_LOCATION gs://DESTINATION_BUCKET_NAME --if-generation-match=GENERATION
```

https://cloud.google.com/storage/docs/request-preconditions#cost_of_preconditions

We would just need to ensure we already have the GENERATION number from when we read the file.
